Add Java 9 Stream APIs

niloc132 opened the pull request:
Partial #9547

niloc132 committed:
Add Java 9 Stream APIs

Partial #9547

niloc132 commented:
https://github.com/niloc132/gwt/actions/runs/5315644262/jobs/9624263435 shows successful build in all three tested Java runtime versions.

niloc132 commented on user/super/com/google/gwt/emul/java/util/stream/Collectors.java:
```diff
@@ -307,6 +337,11 @@ public static <T> Collector<T,?,List<T>> toList() {
     return toCollection(ArrayList::new);
   }
 
+  public static <T> Collector<T, ?, List<T>> toUnmodifiableList() {
```
The unmodifiable methods belong in the java 10 patch, with the tests for these methods.

niloc132 committed:
Remove unmodifiable methods, which belong in a later PR

niloc132 committed:
Remove extra import

niloc132 committed:
Another extra import removed

zbynek commented on user/super/com/google/gwt/emul/java/util/stream/Collectors.java:
```diff
@@ -170,6 +170,34 @@ public static Collector<CharSequence,?,String> joining(CharSequence delimiter) {
+        downstream.supplier(),
+        (A a, T t) -> {
+            mapper.apply(t).forEach(u -> {
+              downstream.accumulator().accept(a, u);
```
from Javadoc

> (If a mapped stream is {@code null} an empty stream is used, instead.)

niloc132 replied:
While, you're right, that's what the Javadoc says... whats the point of treating null as empty here but forcing null to throw elsewhere. I don't care for the inconsistency.

Thanks for pointing this out, I'll make the fix and test for it.

zbynek commented on user/super/com/google/gwt/emul/java/util/stream/DoubleStream.java:
```diff
@@ -239,5 +277,19 @@ static DoubleStream of(double t) {
 
   DoubleSummaryStatistics summaryStatistics();
 
+  default DoubleStream takeWhile(DoublePredicate predicate) {
```
```
DoubleStream.iterate(0.0, d -> d + 2.0)
        .takeWhile(d -> d < 50)
        .boxed()
        .collect(Collectors.toList());
```
should terminate, but with `filter` it won't

niloc132 replied:
Ah thank you, good catch. Ill add the example code you provided to ensure that we stop with infinite streams.

niloc132 replied:
Been unable to work on this for some time, and this next iteration of the work took longer than I had intended, but I think the added tests confirm that I'm on the right track now, PTAL.

I'm going to look into your other note the next chance I get.

zbynek commented on user/super/com/google/gwt/emul/java/util/stream/DoubleStream.java:
```diff
@@ -162,6 +162,30 @@ public boolean tryAdvance(DoubleConsumer action) {
+              return false;
+            }
+            action.accept(next);
+            next = f.applyAsDouble(next);
```
when used together with `limit()` or `takeWhile()`, this calls `f` one more time than needed

niloc132 replied:
Looks like `*Stream.iterate(seed, f)` is also wrong, has been since initial commit. I'll add tests and get this fixed, thanks for calling it out.

niloc132 replied:
Thanks again for calling this out - I updated tests of both old and new `iterate()`, and made the old version delegate to the new one (which will reduce the size impact of using both methods).

niloc132 committed:
"if a mapped stream is null..." there are no items to pass to acc, return early

niloc132 committed:
Test case for null flatmap stream

niloc132 committed:
Add tests for DoubleStream for infinite streams

niloc132 committed:
Fix doublestream impl, correct test off-by-one

niloc132 committed:
Replicate tests to other types

niloc132 committed:
Cleanup before replicating impl

Note especially that this avoids an extra lambda, saving a extra type and
makeFunction call

niloc132 committed:
Replicate implementation, passes tests

niloc132 committed:
Make indentation more consistent with existing code

niloc132 commented on user/super/com/google/gwt/emul/java/util/stream/DoubleStream.java:
```diff
@@ -185,6 +209,39 @@ static DoubleStream of(double t) {
+    Spliterator.OfDouble prev = spliterator();
+    Spliterator.OfDouble spliterator =
+        new Spliterators.AbstractDoubleSpliterator(
+                prev.estimateSize(), prev.characteristics() & ~(Spliterator.SIZED | Spliterator.SUBSIZED)) {
```
Looks like I've should wrap this (and 7 other lines too).

niloc132 committed:
Add tests for old/new iterate functions being called correctly

Also moved newly created test for old behavior to java8 class

niloc132 committed:
Correct Stream impl, to be generalized

niloc132 committed:
Replicate impl to other types, share iterate() implementation

niloc132 committed:
Merge branch 'main' into java9-stream-apis-test

niloc132 committed:
Fix checkstyle line wraps

niloc132 committed:
Merge branch 'main' into java9-stream-apis-test

niloc132 commented:
@zbynek so it seems that Java8 has the same bug that GWT had here previously, so my new test above is failing in legacy dev mode.

```java
import java.util.stream.DoubleStream;
import java.util.Arrays;

class Test {
        public static void main(String[] args) {
                int[] callCount = { 0 };
                double[] values = DoubleStream.iterate(0, val -> {
                        callCount[0]++;
                        return val + 1;
                }).limit(5).toArray();
                
                System.out.println(callCount[0]);
                System.out.println(Arrays.toString(values));
        }
}
```

In Java 11+, this results in
```
4
[0.0, 1.0, 2.0, 3.0, 4.0]
```
But in Java 8 (both my local copy and the build that the GHA runner is using), this yields:
```
5
[0.0, 1.0, 2.0, 3.0, 4.0]
```
Excitingly, Java 8 only has this bug for IntStream, DoubleStream, and LongStream, just plain Stream has the same behavior in 8 and 11/17:
```java
import java.util.stream.Stream;
import java.util.Arrays;

class Test {
        public static void main(String[] args) {
                int[] callCount = { 0 };
                Integer[] values = Stream.iterate(0, val -> {
                        callCount[0]++;
                        return val + 1;
                }).limit(5).toArray(Integer[]::new);
                
                System.out.println(callCount[0]);
                System.out.println(Arrays.toString(values));
        }
}
```
```
4
[0, 1, 2, 3, 4]
```

niloc132 committed:
Exclude tests that fail in Java8 legacy dev mode

Note that these would pass in legacy dev mode in Java 11/17, but we
don't run legacy dev mode tests for those JVMs for a variety of reasons.

niloc132 commented:
https://github.com/niloc132/gwt/actions/runs/6792333097 All current tests pass.
